save: Modifications remain after 'git mv' of annex file

This is a follow-up to gh-3686.  As @loj reported over there, there's an unresolved aspect of that issue.  If a symbolic link is moved with `git mv` and the depth of the link changes, `save` commits the staged changes as is (i.e., no adjustment in the link path) and then git-annex fixes up the path afterwards, leaving the dataset dirty and requiring a second `save` call.  Here's the failing test case linked to from that issue:

```diff
diff --git a/datalad/core/local/tests/test_save.py b/datalad/core/local/tests/test_save.py
index 08b706b857..7384ab0ee7 100644
--- a/datalad/core/local/tests/test_save.py
+++ b/datalad/core/local/tests/test_save.py
@@ -11,6 +11,7 @@
 import os
 import os.path as op
 
+from datalad.cmd import Runner
 from datalad.utils import (
     assure_list,
     Path,
@@ -834,3 +835,17 @@ def test_save_dotfiles():
     for git in [True, False, None]:
         for save_path in [None, "nodot-subdir"]:
             yield check_save_dotfiles, git, save_path
+
+
+@with_tree({"subdir": {"foo": "foocontent"}})
+def test_save_git_mv_fixup(path):
+    ds = Dataset(path).create(force=True)
+    ds.save()
+    assert_repo_status(ds.path)
+    ds.repo.call_git(["mv", op.join("subdir", "foo"), "foo"])
+    ds.save()
+    # Was link adjusted properly?  (gh-3686)
+    eq_(Runner(cwd=ds.path).run(["cat", "foo"])[0],
+        "foocontent")
+    # FIXME: But only the deletion was saved.
+    assert_repo_status(ds.path)
```

<details>
<summary>output</summary>

```
======================================================================
FAIL: datalad.core.local.tests.test_save.test_save_git_mv_fixup
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kyle/src/python/venvs/datalad/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/kyle/src/python/datalad/datalad/tests/utils.py", line 554, in _wrap_with_tree
    return t(*(arg + (d,)), **kw)
  File "/home/kyle/src/python/datalad/datalad/core/local/tests/test_save.py", line 851, in test_save_git_mv_fixup
    assert_repo_status(ds.path)
  File "/home/kyle/src/python/datalad/datalad/tests/utils.py", line 1658, in assert_repo_status
    % (state, state_files, oktobefound))
AssertionError: unexpected content of state "modified": [PosixPath('/tmp/datalad_temp_tree_test_save_git_mv_fixupvj1ers51/foo')] != []
```

</details>

--- a/datalad/core/local/tests/test_save.py
+++ b/datalad/core/local/tests/test_save.py
@@ -949,6 +949,19 @@ def test_save_diff_ignore_submodules_config(path=None):
     assert_repo_status(ds.path)
 
 
+@with_tree({"subdir": {"foo": "foocontent"}})
+def test_save_git_mv_fixup(path=None):
+    ds = Dataset(path).create(force=True)
+    ds.save()
+    assert_repo_status(ds.path)
+    ds.repo.call_git(["mv", op.join("subdir", "foo"), "foo"])
+    ds.save()
+    # Was link adjusted properly?  (gh-3686)
+    assert (ds.pathobj / 'foo').read_text() == "foocontent"
+    # all clean
+    assert_repo_status(ds.path)
+
+
 @with_tree(tree={'somefile': 'file content',
                  'subds': {'file_in_sub': 'other'}})
 def test_save_amend(dspath=None):

--- a/datalad/support/gitrepo.py
+++ b/datalad/support/gitrepo.py
@@ -3336,7 +3336,22 @@ class GitRepo(CoreGitRepo):
         # - commit (with all paths that have been touched, to bypass
         #   potential pre-staged bits)
 
-        need_partial_commit = True if self.get_staged_paths() else False
+        staged_paths = self.get_staged_paths()
+        need_partial_commit = bool(staged_paths)
+        if need_partial_commit and hasattr(self, "call_annex"):
+            # so we have some staged content. let's check which ones
+            # are symlinks -- those could be annex key links that
+            # are broken after a `git-mv` operation
+            # https://github.com/datalad/datalad/issues/4967
+            # call `git-annex pre-commit` on them to rectify this before
+            # saving the wrong symlinks
+            added = status_state['added']
+            tofix = [
+                sp for sp in staged_paths
+                if added.get(self.pathobj / sp, {}).get("type") == "symlink"
+            ]
+            if tofix:
+                self.call_annex(['pre-commit'], files=tofix)
 
         # remove first, because removal of a subds would cause a
         # modification of .gitmodules to be added to the todo list

